Provide TTS voice options, speak and download features

Amnish04 opened the pull request:
After finishing up with all the follow ups filed for TTS, I've extended on the existing functionality as requesting in issue #400 

I have:
1. Updated the **settings** model with an additional property of type `TextToSpeechSettings`
```ts
type TextToSpeechSettings = {
  announceMessages?: boolean;
  voice: TextToSpeechVoices;
};

export enum TextToSpeechVoices {
  ALLOY = "alloy",
  ECHO = "echo",
  FABLE = "fable",
  ONYX = "onyx",
  NOVA = "nova",
  SHIMMER = "shimmer",
}
```
2. Added a **Download as Audio** submenu with a list of **available voices** to choose for download.
![image](https://github.com/tarasglek/chatcraft.org/assets/78865303/8fc0a630-593b-4bc5-a174-e9a426edf6a9)
3. Added an  **Audio Submenu** with 2 options - **Speak** and **Select Voice**.
![image](https://github.com/tarasglek/chatcraft.org/assets/78865303/1bfc1143-4dc7-4ac4-a238-8373c9990edf)
The **Speak** option announces the message content in the voice selected in the **Select Voice** setting. The **Select Voice** setting also applies to the automatic TTS streaming that occurs while the response is being generated.

**Important:** The **Select Voice** menu should ideally be a submenu in the **Option Button** in the prompt form. But since we still have to replace it with the new Menu Component to allow for submenus, I have temporarily placed it in Message Base for now.

This closes #400

Amnish04 commented on src/components/Menu/SubMenu.tsx:
```diff
@@ -3,7 +3,7 @@ import React from "react";
 const SubMenu: React.FC<SubMenuProps> = (props) => {
   return (
-    <ReactMenuSubMenu menuStyle={{ maxHeight: "50vh", overflow: "auto" }} align="center" {...props}>
+    <ReactMenuSubMenu align="center" {...props}>
```
I realized this doesn't work well with **React Menu** as it considers its height to be an aggregate height of self and any nested submenus breaking the nested menus beyond 1 layer.

humphd commented:
I think you should move the voice selection to the Settings vs. this sub-menu.  You're not going to change the voice often enough to need this extra complexity.

Amnish04 commented:
@humphd Makes sense, I've moved the voice selection to Preferences modal, and also added a Preview Audio button.
![image](https://github.com/tarasglek/chatcraft.org/assets/78865303/12dd6220-a705-4067-8b21-e5f036017f2e)

humphd commented on src/components/Message/MessageBase.tsx:
```diff
@@ -316,6 +348,21 @@ function MessageBase({
+
+      const { voice } = settings.textToSpeech;
+
+      // Use lighter tts-1 model when not to minimize latency
```
This comment is confusing, typo?

Amnish04 replied:
Yes, that's a typo. Just pushed the fix

humphd reviewed (changes requested):
Tested this tonight, works amazing!  The code looks great.

Let's remove this sub-menu of voices, since we have it in the settings.  We only need one place to do it.

<img width="524" alt="Screenshot 2024-03-14 at 8 25 22 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/3652bd2d-ac44-4702-81c6-142468d72c63">

Amnish04 commented:
> Let's remove this sub-menu of voices, since we have it in the settings. We only need one place to do it.
> <img alt="Screenshot 2024-03-14 at 8 25 22 PM" width="524" src="https://private-user-images.githubusercontent.com/427398/313029607-3652bd2d-ac44-4702-81c6-142468d72c63.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0NjMzMzksIm5iZiI6MTcxMDQ2MzAzOSwicGF0aCI6Ii80MjczOTgvMzEzMDI5NjA3LTM2NTJiZDJkLWFjNDQtNDcwMi04MWM2LTE0MjQ2OGQ3MmM2My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxNVQwMDM3MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNmRmMTU4MjVhYjZiN2EwYjM4ZGUwM2Q1NTcxNWM3MWFkOTRhYjgxYTFiZDk3YjVjZTNmMzIwMDljODUyNmQwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.n7ELTilD2Pup6ZdSGdfXsu6RuXVWvMqV-qTmbV5g2GI">

I haven't removed the submenu because users might want to download using a specific voice while on the go. In such cases, constantly navigating back and forth between the settings modal could be cumbersome for the user. What do you think?

Amnish04 commented:
@humphd I realized I wasn't handling errors when calling `textToSpeech`. Just made some updates

humphd commented:
> I haven't removed the submenu because users might want to download using a specific voice while on the go. In such cases, constantly navigating back and forth between the settings modal could be cumbersome, especially when they just want to use a different voice for a particular download and not affect the normal setting.
> What do you think?

I think you should remove it :). The reason is that you are too deep into this feature and know too much.  Most people are going to set the voice once, and never think about it again.  They won't switch back and forth all the time, and forcing them to pick a voice when all they want is audio over complicates things.

We have a way to change the voice via Settings, we don't then need to sprinkle that throughout the rest of the UI.

tarasglek commented:
Very cool, I have been wanting this feature forever to generate voice-overs for demos.

I think there is only one case where would wanna switch voice and that's dialog generation, but lets wait for someone to ask us to implement it.

Amnish04 commented:
> I think you should remove it :). The reason is that you are too deep into this feature and know too much. Most people are going to set the voice once, and never think about it again. They won't switch back and forth all the time, and forcing them to pick a voice when all they want is audio over complicates things.
> 
> We have a way to change the voice via Settings, we don't then need to sprinkle that throughout the rest of the UI.

@humphd That also makes sense. Just pushed the changes, now I am using the voice configured in settings for downloads as well

tarasglek commented:
Is there a limit on length of text one can download?

Amnish04 commented:
> Is there a limit on length of spoken audio one can download?

There isn't. I was also thinking about this, but not sure what's the best way to handle.

tarasglek commented:
Can we land this, I wanna use this feature to record youtubes :)

Amnish04 commented:
@humphd Are you also okay with shipping this? I've made the changes you requested

humphd commented:
@Amnish04 absolutely, get this code launched 🚀
